### PR TITLE
Improve enumeration of inferred types in pipeline

### PR DESCRIPTION
**Pull request by MartinGC94**
# PR Summary
Update the type inference for $_ so it uses the same enumeration as the foreach loop.  
Update the native member completion for the *-Object commands to enumerate the types of the input object.  
Update the member completion for foreach-object to not add the parenthesis when completing methods.
Update the type inference for Where, Sort and Select so the inferred type after these commands are enumerated.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 [hunk 1]
```diff
@@ -571,6 +571,19 @@ ConstructorTestClass(int i, bool b)
         $res.CompletionMatches[0].CompletionText | Should -BeExactly '-Directory'
     }
 
+    It 'Should complete method name without starting parenthesis in foreach-object' {
+        $TestString = '1..10 | ForEach-Object gettype'
+        $res = TabExpansion2 -inputScript $TestString
+        $res.CompletionMatches[0].CompletionText | Should -BeExactly GetType
```
Comment (iSazonov) [inline test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:577]:
```suggestion
        $res.CompletionMatches[0].CompletionText | Should -BeExactly 'GetType'
```

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 [hunk 2]
```diff
@@ -571,6 +571,19 @@ ConstructorTestClass(int i, bool b)
...
+    }
+
+    It 'Should enumerate types when completing member names for Select-Object' {
+        $TestString = '"Hello","World" | select-object '
+        $res = TabExpansion2 -inputScript $TestString
+        $res | Should -HaveCount 1
+        $res.CompletionMatches[0].CompletionText | Should -BeExactly Length
```
Comment (iSazonov) [inline test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:584]:
```suggestion
        $res.CompletionMatches[0].CompletionText | Should -BeExactly 'Length'
```

**Comment by iSazonov** [comment]
@MartinGC94 Please look CI fails.

**Review by iSazonov — APPROVED** [review]

**Comment by MartinGC94** [comment]
@iSazonov Sorry, one more minor change so this works as expected:
```
$ThisIsAString = [System.IO.Directory]::EnumerateDirectories("C:\") | Select-Object -First 1
$ThisIsAString.<Tab>
```

**Review by adityapatwardhan — APPROVED** [review]
Signed off as maintainer

**Comment by adityapatwardhan** [comment]
@daxian-dbw - I would like your review as a SME.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 3]
```diff
@@ -6077,7 +6088,7 @@ internal static void CompleteMemberByInferredType(TypeInferenceContext context,
             }
         }
 
-        private static void AddInferredMember(object member, WildcardPattern memberNamePattern, List<CompletionResult> results, HashSet<string> excludedMembers)
+        private static void AddInferredMember(object member, WildcardPattern memberNamePattern, List<CompletionResult> results, HashSet<string> excludedMembers, bool addMethodparenthesis)
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:6091]:
```suggestion
        private static void AddInferredMember(object member, WildcardPattern memberNamePattern, List<CompletionResult> results, HashSet<string> excludedMembers, bool addMethodParenthesis)
```

**Review by daxian-dbw — APPROVED** [review]
LGTM except that the parameter name should use camel case.

**Comment by daxian-dbw** [comment]
@MartinGC94 You will need to do additional changes on the uses of that parameter within different methods. My suggested changes above didn't include those.

**Pull request merged**